### PR TITLE
Vendor-fetch download to lib directory specified in project.json (#1422)

### PR DESCRIPTION
**Pull request by brian-sinquin**
This PR refers to the issue #1422 
* [x] vendor-fetch should download to `dependency-search-path` field set in `project.json` by default
* [x] vendor-fetch should add libraries in the target library list in `project.json` (~~with an option to specify a target~~)
    * [ ] Proper JSON file update (`dependencies` field of root or specified target) after fetching
    * [ ] Create `dependencies` field if it does not exist
* [ ] Fetch missing libraries that are specified in `project.json`

### CODE REVIEW

**Comment by brian-sinquin** [comment]
For now, i worked on :

* The default vendor-fetch download path is set using the first dependency-path specified in the project JSON file.
* This behavior is skipped if --path is specified (vendor-fetch in the specified path)
* Adds the missing libraries in the global dependency entry of the project JSON file

I think i need a review at that point and think about refining the code, and making it more versatile.

**Comment by lerno** [comment]
Just let me know when you need the review.

**Comment by brian-sinquin** [comment]
Test should be running now, i fixed the case where my code does not find the project JSON file. Please review when you can, then i will add the fetching of missing dependencies. Also, the json API works well for reading but is lacks some helper functions for editing.

**Comment by lerno** [comment]
It looks like there are some issues to address

**Comment by brian-sinquin** [comment]
> It looks like there are some issues to address

I manage to compile well on my computer, i updated const types that were not before.
I face an edge case in my JSON project file edition (to add fetched libraries), all empty arrays of the file are modified while i want to only alter the `dependencies` entry. Something may wring in my json object manipulation but i dont see where.

**Comment by lerno** [comment]
Things that have to be fixed for it to compile on all compilers:
```
D:\a\c3c\c3c\src\compiler\compiler.c:878:20: error: result of comparison against a string literal is unspecified (use an explicit string comparison function instead) [-Werror,-Wstring-compare]
  878 |         if (options->path == DEFAULT_PATH)
      |                           ^  ~~~~~~~~~~~~
D:\a\c3c\c3c\src\compiler\compiler.c:907:4: error: assigning to 'char *' from 'const char *' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
  907 |                         vec_add(fetched_libraries, lib);
      |                         ^                          ~~~
D:\a\c3c\c3c\src\compiler/../utils/lib.h:339:29: note: expanded from macro 'vec_add'
  339 |         (vec_)[vec_size(vec_) - 1] = value_; \
      |                                    ^ ~~~~~~
D:\a\c3c\c3c\src\compiler\compiler.c:917:32: error: passing 'char **' to parameter of type 'const char **' discards qualifiers in nested pointer types [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
  917 |         add_libraries_to_project_file(fetched_libraries, options->project_options.target_name);
      |                                       ^~~~~~~~~~~~~~~~~
D:\a\c3c\c3c\src\compiler/../build/project.h:6:49: note: passing argument to parameter 'libs' here
    6 | void add_libraries_to_project_file(const char** libs, const char* target_name);
      |        
```

**Comment by lerno** [comment]
One CI doesn't work because vendor-fetch now requires an argument:
```
Run build\Release\c3c.exe vendor-fetch raylib5
  build\Release\c3c.exe vendor-fetch raylib5
  shell: C:\Windows\system32\cmd.EXE /D /E:ON /V:OFF /S /C "CALL "{0}""
  env:
    LLVM_RELEASE_VERSION_WINDOWS: 18
    LLVM_RELEASE_VERSION_MAC: 17
    LLVM_RELEASE_VERSION_LINUX: 17
    LLVM_RELEASE_VERSION_UBUNTU20: 17
    LLVM_DEV_VERSION: 20
No project file found, use --path to specify custom destination.
```
But this looks wrong: it should be possible to use vendor-fetch without a project?

**Comment by brian-sinquin** [comment]
> One CI doesn't work because vendor-fetch now requires an argument:
> 
> ```
> Run build\Release\c3c.exe vendor-fetch raylib5
>   build\Release\c3c.exe vendor-fetch raylib5
>   shell: C:\Windows\system32\cmd.EXE /D /E:ON /V:OFF /S /C "CALL "{0}""
>   env:
>     LLVM_RELEASE_VERSION_WINDOWS: 18
>     LLVM_RELEASE_VERSION_MAC: 17
>     LLVM_RELEASE_VERSION_LINUX: 17
>     LLVM_RELEASE_VERSION_UBUNTU20: 17
>     LLVM_DEV_VERSION: 20
> No project file found, use --path to specify custom destination.
> ```
> 
> But this looks wrong: it should be possible to use vendor-fetch without a project?

You are right. It is fixed, it will download in the default path (cwd) if it does not find a project file.

File: src/compiler/compiler.c [hunk 1]
```diff
@@ -861,7 +861,7 @@ const char * vendor_fetch_single(const char* lib, const char* path)
 void vendor_fetch(BuildOptions *options)
 {
 	unsigned count = 0;
-	if (options->path == DEFAULT_PATH)
+	if (strcmp(options->path, DEFAULT_PATH) == 0)
```
Comment (lerno) [inline src/compiler/compiler.c:864]: There is a str_eq which just returns a bool, could you use that one instead?

Comment (brian-sinquin) [reply src/compiler/compiler.c]: done.

Comment (lerno) [reply src/compiler/compiler.c]: It's not a show-stopper though. I can do it later.

**Comment by brian-sinquin** [comment]
So far i added the two first features of the initial PR.

Missing features : 
- Fetch a library that is needed but not present in the search path
- Add fetched libraries in the `dependencies` member of a specified target (JSON root for the moment)

Know bugs/edge cases :

- No checks / correction if the "depencies", "dependencies-search-path" ... do not exists
- Something is wrong with my JSON editing code (populating wrong array entries if they are empty)

**Comment by lerno** [comment]
The JSON bug seems worrying?

**Comment by brian-sinquin** [comment]
> The JSON bug seems worrying?

Let's say you begin with no search-path : 
```json
{
  "langrev": "1",
  "warnings":  [ "no-unused" ],
  "dependency-search-paths":  [ ],
  "dependencies":  [ ],
  "authors":  [ "John Doe <john.doe@example.com>" ],
  "version": "0.1.0",
  "sources":  [ "src/**" ],
  "output": "build",
  "targets": {
    "myAAAGame": {
      "type": "executable",
    },
  },
  "cpu": "generic",
  "opt": "O0",
}
```

After `c3c vendor-fetch raylib5` you will end up with : 
```json
{
  "langrev": "1",
  "warnings":  [ "no-unused" ],
  "dependency-search-paths":  [ "raylib5" ],
  "dependencies":  [ "raylib" ],
  "authors":  [ "John Doe <john.doe@example.com>" ],
  "version": "0.1.0",
  "sources":  [ "src/**" ],
  "output": "build",
  "targets": {
    "myAAAGame": {
      "type": "executable",
    },
  },
  "cpu": "generic",
  "opt": "O0",
}
```
The compilator will search for `./raylib5/raylib5.c3l` while it is actualy at `./raylib5/raylib5.c3l`.
I dont know why the `dependencies-search` gets populated, i may not use the JSONObject properly.

**Comment by lerno** [comment]
Ok I know why. First of all, the JSON object were never intended to be manipulated – to just be read-only structure. This creates lots of problems when it's mutated. In this case, an empty array is reused. So you're actually reusing the single empty array in the JSON structure!

**Comment by lerno** [comment]
The JSON stuff really needs to be brought in to use the general C3 allocator.

**Comment by lerno** [comment]
Thank you!

**Comment by brian-sinquin** [comment]
> Ok I know why. First of all, the JSON object were never intended to be manipulated – to just be read-only structure. This creates lots of problems when it's mutated. In this case, an empty array is reused. So you're actually reusing the single empty array in the JSON structure!

Yeah, I saw it was more a read API but I tried to use the abstract object. Thank you for the fix

**Comment by lerno** [comment]
I updated it further, it should be straightforward to manipulate now.

**Pull request merged**